internal/notebook/resolver.go

```diff
@@ -0,0 +1,132 @@
+	editor   parserv1.ParserServiceServer
+}
+
+func NewNotebookResolver(notebook *parserv1.Notebook) *NotebookResolver {
+	return &NotebookResolver{
+		notebook: notebook,
+		editor:   editorservice.NewParserServiceServer(zap.NewNop()),
+	}
+}
+
+func (r *NotebookResolver) ParseNotebook(context context.Context) (*parserv1.Notebook, error) {
+	// make id sticky only for resolving purposes
+	for _, cell := range r.notebook.Cells {
+		if cell.GetKind() != parserv1.CellKind_CELL_KIND_CODE {
+			continue
+		}
+
+		_, ok := cell.Metadata["id"]
+		if ok {
+			continue
+		}
+
+		if cell.Metadata == nil {
+			return nil, fmt.Errorf("cell metadata is missing")
+		}
+
+		cell.Metadata["id"] = cell.Metadata["runme.dev/id"]
+	}
+
+	// properly parse frontmatter and notebook/cell metadata
+	ser, err := r.editor.Serialize(context, &parserv1.SerializeRequest{Notebook: r.notebook})
+	if err != nil {
+		return nil, err
+	}
+	des, err := r.editor.Deserialize(context, &parserv1.DeserializeRequest{Source: ser.Result})
+	if err != nil {
+		return nil, err
+	}
+
+	return des.Notebook, nil
+}
+
+func (r *NotebookResolver) ResolveDaggerShell(context context.Context, cellIndex uint32) (*notebookv1alpha1.ResolveNotebookResponse, error) {
+	notebook, err := r.ParseNotebook(context)
+	if err != nil {
+		return nil, err
+	}
+
+	var targetCell *parserv1.Cell
+	targetName := ""
+	for idx, cell := range notebook.Cells {
+		if idx != int(cellIndex) {
+			continue
+		}
+
+		id, okID := cell.Metadata["runme.dev/id"]
+		known, okKnown := cell.Metadata["name"]
+		generated := cell.Metadata["runme.dev/nameGenerated"]
+		if !okID && !okKnown {
+			continue
+		}
+
+		isGenerated, err := strconv.ParseBool(generated)
+		if !okKnown || isGenerated || err != nil {
+			known = fmt.Sprintf("DAGGER_%s", id)
+		}
+
+		targetCell = cell
+		targetName = known
+		break
+	}
+
+	if notebook.Frontmatter == nil || !strings.Contains(strings.Trim(notebook.Frontmatter.Shell, " \t\r\n"), "dagger shell") {
+		return &notebookv1alpha1.ResolveNotebookResponse{Script: targetCell.GetValue()}, nil
+	}
+
+	script := daggershell.NewScript()
+	for _, cell := range notebook.Cells {
+		if cell.GetKind() != parserv1.CellKind_CELL_KIND_CODE {
+			continue
+		}
+
+		languageID := cell.GetLanguageId()
+		if languageID != "sh" && languageID != "dagger" {
+			continue
+		}
+
+		id, okID := cell.Metadata["runme.dev/id"]
+		known, okName := cell.Metadata["runme.dev/name"]
+		generated := cell.Metadata["runme.dev/nameGenerated"]
+		if !okID && !okName {
+			continue
+		}
+
+		isGenerated, err := strconv.ParseBool(generated)
+		if !okName || isGenerated || err != nil {
+			known = fmt.Sprintf("DAGGER_%s", id)
+		}
+
+		snippet := cell.GetValue()
+		if err := script.DeclareFunc(known, snippet); err != nil {
+			return nil, err
+		}
+	}
+
+	var rendered bytes.Buffer
+	if err := script.RenderWithCall(&rendered, targetName); err != nil {
+		return nil, err
+	}
+
+	return &notebookv1alpha1.ResolveNotebookResponse{
```

Comment:
Suggestion: Returning `renderer.String()` should be easier to reason about and tests. The service can wrap it as it pleases without polluting other files with `notebookv1alpha1`.

Reply:
Good point!